error: use of undeclared identifier 'isunordered'; did you mean 'std::isunordered'?

env:
===
ubuntu 16.04
clang++ 10.01

step:
===
cmake -GNinja -B ../iree-build/ -S . \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DIREE_ENABLE_ASSERTIONS=ON \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DIREE_ENABLE_LLD=ON

cmake --build ../iree-build/

error:
===
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:340:11: error: use of undeclared identifier 'isunordered'; did you mean 'std::isunordered'?
  return (isunordered(lhs, rhs) || (lhs == rhs)) ? 1 : 0;
          ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:829:3: note: 'std::isunordered' declared here
  isunordered(float __x, float __y)

> I can't see which source fine tha

davidcheng@davidcheng-System-Product-Name:~/workspace/repo/github/iree$ cmake --build ../iree-build/
[0/2] Re-checking globbed directories...
[962/4680] Building CXX object third_party/llvm-project/llvm/lib/Analysis/CMakeFiles/LLVMAnalysis.dir/LoopAccessAnalysis.cpp.o
[4669/4680] Building CXX object iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_module_test.dir/module_test.cc.o
FAILED: iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_module_test.dir/module_test.cc.o
/home/davidcheng/workspace/repo/github/clang+llvm-10.0.1-x86_64-linux-gnu-ubuntu-16.04/bin/clang++ -DIREE_HAVE_EMITC_DIALECT -isystem /home/davidcheng/workspace/repo/github/iree -isystem /home/davidcheng/workspace/repo/github/iree-build -isystem /home/davidcheng/workspace/repo/github/iree/third_party/googletest/googlemock/include -isystem /home/davidcheng/workspace/repo/github/iree/third_party/googletest/googlemock -isystem /home/davidcheng/workspace/repo/github/iree/third_party/googletest/googletest/include -isystem /home/davidcheng/workspace/repo/github/iree/third_party/googletest/googletest -O2 -g  -fPIE -fvisibility=hidden -fno-rtti -fno-exceptions -Werror -Wall -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-deprecated-declarations -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -std=gnu++14 -MD -MT iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_module_test.dir/module_test.cc.o -MF iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_module_test.dir/module_test.cc.o.d -o iree/vm/test/emitc/CMakeFiles/iree_vm_test_emitc_module_test.dir/module_test.cc.o -c /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:340:11: error: use of undeclared identifier 'isunordered'; did you mean 'std::isunordered'?
  return (isunordered(lhs, rhs) || (lhs == rhs)) ? 1 : 0;
          ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:829:3: note: 'std::isunordered' declared here
  isunordered(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:346:11: error: use of undeclared identifier 'isunordered'; did you mean 'std::isunordered'?
  return (isunordered(lhs, rhs) || (lhs != rhs)) ? 1 : 0;
          ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:829:3: note: 'std::isunordered' declared here
  isunordered(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:349:10: error: use of undeclared identifier 'isless'; did you mean 'std::isless'?
  return isless(lhs, rhs) ? 1 : 0;
         ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:751:3: note: 'std::isless' declared here
  isless(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:352:11: error: use of undeclared identifier 'isunordered'; did you mean 'std::isunordered'?
  return (isunordered(lhs, rhs) || isless(lhs, rhs)) ? 1 : 0;
          ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:829:3: note: 'std::isunordered' declared here
  isunordered(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:352:36: error: use of undeclared identifier 'isless'; did you mean 'std::isless'?
  return (isunordered(lhs, rhs) || isless(lhs, rhs)) ? 1 : 0;
                                   ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:751:3: note: 'std::isless' declared here
  isless(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:355:10: error: use of undeclared identifier 'islessequal'; did you mean 'std::islessequal'?
  return islessequal(lhs, rhs) ? 1 : 0;
         ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:777:3: note: 'std::islessequal' declared here
  islessequal(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:358:11: error: use of undeclared identifier 'isunordered'; did you mean 'std::isunordered'?
  return (isunordered(lhs, rhs) || islessequal(lhs, rhs)) ? 1 : 0;
          ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:829:3: note: 'std::isunordered' declared here
  isunordered(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:358:36: error: use of undeclared identifier 'islessequal'; did you mean 'std::islessequal'?
  return (isunordered(lhs, rhs) || islessequal(lhs, rhs)) ? 1 : 0;
                                   ^
/usr/lib/gcc/x86_64-linux-gnu/5.5.0/../../../../include/c++/5.5.0/cmath:777:3: note: 'std::islessequal' declared here
  islessequal(float __x, float __y)
  ^
In file included from /home/davidcheng/workspace/repo/github/iree/iree/vm/test/emitc/module_test.cc:11:
In file included from /home/davidcheng/workspace/repo/github/iree-build/iree/vm/test/emitc/arithmetic_ops.h:2:
/home/davidcheng/workspace/repo/github/iree/iree/vm/ops.h:361:10: error: use of undeclared identifier 'isnan'
  return isnan(operand) ? 1 : 0;
         ^
9 errors generated.
[4674/4680] Building CXX object iree/vm/CMakeFiles/iree_vm_bytecode_module_test.dir/bytecode_dispatch_test.cc.o
ninja: build stopped: subcommand failed.

--- a/iree/vm/test/emitc/module_test.cc
+++ b/iree/vm/test/emitc/module_test.cc
@@ -4,6 +4,13 @@
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+// TODO: We should not be including C implementation-only headers in a C++
+// module like this. In order to make this work for the moment across
+// runtime libraries that are strict, do a global using of the std namespace.
+// See #7605
+#include <cmath>
+using namespace std;
+
 #include "iree/base/logging.h"
 #include "iree/base/status_cc.h"
 #include "iree/testing/gtest.h"

